models/middleware.js

```diff
@@ -0,0 +1,89 @@
+var app = require('../app');
+var loopback = require('loopback');
+var debug = require('debug')('workspace:middleware');
+var ConfigFile = app.models.ConfigFile;
+var path = require('path');
+
+/**
+ * Defines a `Middleware` configuration.
+ * @class Middleware
+ * @inherits Definition
+ */
+
+var Middleware = app.models.Middleware;
+
+/**
+ * - `facetName` is required and must refer to an existing facet
+ *
+ * @header Property Validation
+ */
+Middleware.validatesPresenceOf('facetName');
+
+/**
+ * Serialize the middleware model instances to the JSON object for
+ * middleware.json
+ * @param {*[]} cache The cache data source
+ * @param {String} facetName Facet name
+ * @returns {ConfigFile}
+ */
+Middleware.serialize = function(cache, facetName) {
+  var middlewarePath = path.join(facetName, 'middleware.json');
+  var instances = Middleware.allFromCache(cache);
+
+  var middlewareConfig = {};
+  instances.forEach(function(middlewareDef) {
+    if (middlewareDef.facetName === facetName) {
+      debug('Serializing middleware %j', middlewareDef);
+      var phaseDef = middlewareConfig[middlewareDef.phase];
+      if (!phaseDef) {
+        phaseDef = {};
+        middlewareConfig[middlewareDef.phase] = phaseDef;
+      }
+      if (middlewareDef.name && middlewareDef.name[0] !== '*') {
+        phaseDef[middlewareDef.name] =
+          Middleware.getConfigFromData(middlewareDef);
+      }
+    }
+  });
+  debug('middleware.json: %j', middlewareConfig);
+  return new ConfigFile({
+    path: middlewarePath,
+    data: middlewareConfig
+  });
+}
+
+/**
+ * Load the middleware config from the file into cache
+ * @param cache
+ * @param facetName
+ * @param configFile
+ */
+Middleware.deserialize = function(cache, facetName, configFile) {
+  var middlewareDefs = configFile.data || {};
+
+  var phases = Object.keys(middlewareDefs);
+
+  phases.forEach(function(phase) {
+    var defs = middlewareDefs[phase];
+    var def = null;
+    for (var d in defs) {
+      def = defs[d];
+      def.configFile = configFile.path;
+      def.phase = phase;
+      def.facetName = facetName;
+      def.name = d;
+      debug('loading [%s] middleware into cache', def.name);
+      Middleware.addToCache(cache, def);
+    }
+    if (!def) {
+      // Keep the phase information by adding an empty middleware config
```

Comment:
AFAICT, there is not unit-test to verify how this special entry is handled. Could you please add one?